mm2src/coins/sia/src/transaction.rs

```diff
@@ -0,0 +1,1587 @@
+    where
+        S: Serializer,
+    {
+        serializer.serialize_str(&self.to_u128().to_string())
+    }
+}
+
+impl Currency {
+    pub fn new(lo: u64, hi: u64) -> Self { Currency { lo, hi } }
+
+    pub fn to_u128(&self) -> u128 { ((self.hi as u128) << 64) | (self.lo as u128) }
+}
+
+impl From<u64> for Currency {
+    fn from(value: u64) -> Self { Currency { lo: value, hi: 0 } }
+}
+
+// Currency remains the same data structure between V1 and V2 however the encoding changes
+#[derive(Clone, Debug)]
+pub enum CurrencyVersion<'a> {
+    V1(&'a Currency),
+    V2(&'a Currency),
+}
+
+impl<'a> Encodable for CurrencyVersion<'a> {
+    fn encode(&self, encoder: &mut Encoder) {
+        match self {
+            CurrencyVersion::V1(currency) => {
+                let mut buffer = [0u8; 16];
+
+                buffer[8..].copy_from_slice(&currency.lo.to_be_bytes());
+                buffer[..8].copy_from_slice(&currency.hi.to_be_bytes());
+
+                // Trim leading zero bytes from the buffer
+                let trimmed_buf = match buffer.iter().position(|&x| x != 0) {
+                    Some(index) => &buffer[index..],
+                    None => &buffer[..], // In case all bytes are zero
+                };
+                encoder.write_len_prefixed_bytes(trimmed_buf);
+            },
+            CurrencyVersion::V2(currency) => {
+                encoder.write_u64(currency.lo);
+                encoder.write_u64(currency.hi);
+            },
+        }
+    }
+}
+
+#[serde_as]
+#[derive(Clone, Debug, Deserialize, Serialize, PartialEq)]
+#[serde(deny_unknown_fields)]
+pub struct SatisfiedPolicy {
+    #[serde_as(as = "FromInto<SpendPolicyHelper>")]
+    pub policy: SpendPolicy,
+    #[serde_as(as = "Vec<FromInto<PrefixedSignature>>")]
+    #[serde(default, skip_serializing_if = "Vec::is_empty")]
+    pub signatures: Vec<Signature>,
+    #[serde(default, skip_serializing_if = "Vec::is_empty")]
+    pub preimages: Vec<Vec<u8>>,
+}
+
+impl Encodable for Signature {
+    fn encode(&self, encoder: &mut Encoder) { encoder.write_slice(&self.to_bytes()); }
+}
+
+impl Encodable for SatisfiedPolicy {
+    fn encode(&self, encoder: &mut Encoder) {
+        self.policy.encode(encoder);
+        let mut sigi: usize = 0;
+        let mut prei: usize = 0;
+
+        fn rec(policy: &SpendPolicy, encoder: &mut Encoder, sigi: &mut usize, prei: &mut usize, sp: &SatisfiedPolicy) {
+            match policy {
+                SpendPolicy::PublicKey(_) => {
+                    if *sigi < sp.signatures.len() {
+                        sp.signatures[*sigi].encode(encoder);
+                        *sigi += 1;
+                    } else {
+                        // Sia Go code panics here but our code assumes encoding will always be successful
+                        // TODO: check if Sia Go will fix this
+                        encoder.write_string("Broken PublicKey encoding, see SatisfiedPolicy::encode")
+                    }
+                },
```

Comment:
So, for `SpendPolicy::PublicKey` we will always have one signature? should we increment the `sigi` index elsewhere to make this clearer?
I believe this recursive function can be written much more better. We should pass a signature to an `encode_public_key` function, same thing with preimage and `SpendPolicy::Hash`. Incrementing can be done inside `SpendPolicy::Threshold` and `SpendPolicy::UnlockConditions` handling.

P.S. it's up to you in the end how you want to write this, but maybe we shouldn't copy the exact logic written by the sia library to make it easier for future devs to extend this rust library.

Reply:
>SpendPolicy::PublicKey we will always have one signature?

The data type itself would allow for additional signatures, so the concern would be developer error or a malicious party intentionally sending a "corrupt" SpendPolicy. We would need prechecks on SpendPolicy objects prior to encoding if they are sourced from anywhere other than our client. 

This logic can be considered as "consensus code" because it deals with the encoding used to create signatures. The encoding must be the same. The Sia devs insist their code is fine because it would require developer error to reach this case.

Trying to find a better solution now. I am certainly open to suggestions. I can ask that they make this case return Null or 00s or anything other reliably reproducible pattern.

>We should pass a signature to an encode_public_key function, same thing with preimage and SpendPolicy::Hash. Incrementing can be done inside SpendPolicy::Threshold and SpendPolicy::UnlockConditions handling.

I need clarification on this. Could you provide psuedocode or similar?

Reply:
> I need clarification on this. Could you provide psuedocode or similar?

I realise now that this is not straightforward, a better approach is to follow a stack-based pattern as below

```rust
impl Encodable for SatisfiedPolicy {
    fn encode(&self, encoder: &mut Encoder) {
        self.policy.encode(encoder);

        let mut signature_iter = self.signatures.iter();
        let mut preimage_iter = self.preimages.iter();
        let mut stack = vec![&self.policy];

        while let Some(policy) = stack.pop() {
            match policy {
                SpendPolicy::PublicKey(_) => {
                    if let Some(signature) = signature_iter.next() {
                        signature.encode(encoder);
                    } else {
                        encoder.write_string("Broken PublicKey encoding, see SatisfiedPolicy::encode");
                    }
                },
                SpendPolicy::Hash(_) => {
                    if let Some(preimage) = preimage_iter.next() {
                        encoder.write_len_prefixed_bytes(preimage);
                    } else {
                        encoder.write_string("Broken Hash encoding, see SatisfiedPolicy::encode");
                    }
                },
                SpendPolicy::Threshold { n: _, of } => {
                    // Push sub-policies in reverse order to ensure they are processed in the correct order
                    for sub_policy in of.iter().rev() {
                        stack.push(sub_policy);
                    }
                },
                SpendPolicy::UnlockConditions(uc) => {
                    for unlock_key in &uc.unlock_keys {
                        if let UnlockKey::Ed25519(public_key) = unlock_key {
                            stack.push(&SpendPolicy::PublicKey(*public_key));
                        }
                        // else FIXME consider when this is possible, is it always developer error or could it be forced maliciously?
                    }
                },
                _ => {},
            }
        }
    }
}
```

I think this is much simpler than index incrementing

Reply:
These are not semantically equivalent. The stack-based approach is a BFS, whereas the recursive approach is a DFS. So if you have nested `thresh` policies, the two approaches will encode the signatures/preimages in a different order.

I can understand the aversion to raw array indices, so my suggestion would be to use iterators in the recursive approach.

Reply:
btw, the reason we use recursion for encoding policies is that policies are themselves a recursive data structure, and they are validated recursively. But the signatures and preimages are not "inlined" into the same recursive structure -- they are provided "out-of-band" in separate arrays. This forces us to do slightly gross things like increment an index (or mutate an iterator...) from within a recursive function.

An alternative approach would be to construct a new recursive type where the validation data lives alongside its corresponding policy; something like:
```go
type SatisfiedPolicy interface {
    Policy() SpendPolicy
    Data() any
}

type SatisfiedPublicKeyPolicy struct {
    pk  types.PublicKey
    sig types.Signature
}
func (spk SatisfiedPublicKeyPolicy) Policy() SpendPolicy { return PolicyPublicKey(spk.pk) }
func (spk SatisfiedPublicKeyPolicy) Data() any { return spk.sig }
// etc.
```
However, this ends up being messy in different ways (e.g. it requires defining new "satisfied" variant types for every policy), so I decided against it

Reply:
> These are not semantically equivalent. The stack-based approach is a BFS, whereas the recursive approach is a DFS. So if you have nested thresh policies, the two approaches will encode the signatures/preimages in a different order.

True, bad mistake from my side where I confused pop/push as popping/pushing from/to index 0, I assume the below is DFS
```rust
impl Encodable for SatisfiedPolicy {
    fn encode(&self, encoder: &mut Encoder) {
        self.policy.encode(encoder);

        let mut signature_iter = self.signatures.iter();
        let mut preimage_iter = self.preimages.iter();
        let mut stack = VecDeque::new();
        stack.push_back(&self.policy);

        while let Some(policy) = stack.pop_front() {
            match policy {
                SpendPolicy::PublicKey(_) => {
                    if let Some(signature) = signature_iter.next() {
                        signature.encode(encoder);
                    } else {
                        encoder.write_string("Broken PublicKey encoding, see SatisfiedPolicy::encode");
                    }
                },
                SpendPolicy::Hash(_) => {
                    if let Some(preimage) = preimage_iter.next() {
                        encoder.write_len_prefixed_bytes(preimage);
                    } else {
                        encoder.write_string("Broken Hash encoding, see SatisfiedPolicy::encode");
                    }
                },
                SpendPolicy::Threshold { n: _, of } => {
                    for sub_policy in of.iter().rev() {
                        stack.push_front(sub_policy);
                    }
                },
                SpendPolicy::UnlockConditions(uc) => {
                    for unlock_key in &uc.unlock_keys {
                        if let UnlockKey::Ed25519(public_key) = unlock_key {
                            stack.push_front(&SpendPolicy::PublicKey(*public_key));
                        }
                        // else FIXME consider when this is possible, is it always developer error or could it be forced maliciously?
                    }
                },
                _ => {},
            }
        }
    }
}
```